### PR TITLE
[15.0][FIX] Duplicate some mail.thread.phone code in res.partner

### LINKED ISSUE #79460: [ODOO15] AttributeError: 'res.partner' object has no attribute 'phone_get_sanitized_number'
Impacted versions:
commit 3e6751a8009594047c17b9bd231b9f78ef987a3d (HEAD -> 15.0, origin/15.0)

Steps to reproduce:

```sh
python odoo/odoo-bin -c odoo15.conf 
```

```odoo15.conf```

```conf
[options]
addons_path = /sdl/odoo/addons,/sdl/odoo/odoo/addons
admin_passwd = 
data_dir = /sdl/.data/odoo15
db_host = 172.19.0.2
db_name = odoo15-local
db_port = 5432
db_sslmode = disable
db_user = odoo
db_password = odoo
email_from = False
log_level = info
max_cron_threads = 1
proxy_mode = True
# workers = 2
limit_time_cpu = 600
limit_time_real = 1200
#test_enable = False
#test_file = False
#test_tags = sdl
translate_modules = ['all']
without_demo = False
smtp_password = False
smtp_port = 25
smtp_server = localhost
smtp_ssl = False
smtp_user = False
```

Current behavior:
```log
2021-11-08 05:47:30,808 28289 WARNING odoo15-local odoo.modules.loading: Module crm demo data failed to install, installed without demo data 
Traceback (most recent call last):
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/tools/convert.py", line 680, in _tag_root
    f(rec)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/tools/convert.py", line 333, in _tag_function
    _eval_xml(self, rec, env)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/tools/convert.py", line 204, in _eval_xml
    return odoo.api.call_kw(model, method_name, args, kwargs)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/api.py", line 461, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/api.py", line 448, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 895, in action_set_lost
    res = self.action_archive()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 5046, in action_archive
    return self.filtered(lambda record: record[self._active_name]).toggle_active()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 883, in toggle_active
    res = super(Lead, self).toggle_active()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/mail/models/mail_activity_mixin.py", line 337, in toggle_active
    return super(MailActivityMixin, self).toggle_active()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 5039, in toggle_active
    active_recs[self._active_name] = False
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 5875, in __setitem__
    return self._fields[key].__set__(self, value)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1217, in __set__
    records.write({self.name: write_value})
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 683, in write
    self._handle_won_lost(vals)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 801, in _handle_won_lost
    leads_reach_lost._pls_increment_frequencies(to_state='lost')
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 1995, in _pls_increment_frequencies
    new_frequencies_by_team, existing_frequencies_by_team = self._pls_prepare_update_frequency_table(target_state=from_state or to_state)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 2153, in _pls_prepare_update_frequency_table
    leads_values_dict = pls_leads._pls_get_lead_pls_values(domain=domain)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 2373, in _pls_get_lead_pls_values
    self.flush(['probability'])
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 5628, in flush
    self.recompute(fnames, records=records)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 6108, in recompute
    process(field)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 6079, in process
    field.recompute(recs)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1243, in recompute
    self.compute_value(recs)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1265, in compute_value
    records._compute_field_value(self)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/mail/models/mail_thread.py", line 410, in _compute_field_value
    return super()._compute_field_value(field)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 4248, in _compute_field_value
    getattr(self, field.compute)()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 451, in _compute_probabilities
    lead_probabilities = self._pls_get_naive_bayes_probabilities()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 1873, in _pls_get_naive_bayes_probabilities
    leads_values_dict = self._pls_get_lead_pls_values(domain=domain)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 2418, in _pls_get_lead_pls_values
    value = lead[field].id if isinstance(lead[field], models.BaseModel) else lead[field]
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 5866, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1054, in __get__
    self.recompute(record)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1243, in recompute
    self.compute_value(recs)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1265, in compute_value
    records._compute_field_value(self)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/mail/models/mail_thread.py", line 410, in _compute_field_value
    return super()._compute_field_value(field)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 4248, in _compute_field_value
    getattr(self, field.compute)()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 421, in _compute_phone_state
    if lead.phone:
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1054, in __get__
    self.recompute(record)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1243, in recompute
    self.compute_value(recs)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1265, in compute_value
    records._compute_field_value(self)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/mail/models/mail_thread.py", line 410, in _compute_field_value
    return super()._compute_field_value(field)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 4248, in _compute_field_value
    getattr(self, field.compute)()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 409, in _compute_phone
    if lead.partner_id.phone and lead._get_partner_phone_update():
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 637, in _get_partner_phone_update
    partner_phone_formatted = self.partner_id.phone_get_sanitized_number(number_fname='phone') or self.partner_id.phone or False
AttributeError: 'res.partner' object has no attribute 'phone_get_sanitized_number'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/modules/loading.py", line 86, in load_demo
    load_data(cr, idref, mode, kind='demo', package=package)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/modules/loading.py", line 69, in load_data
    tools.convert_file(cr, package.name, filename, idref, mode, noupdate, kind)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/tools/convert.py", line 745, in convert_file
    convert_xml_import(cr, module, fp, idref, mode, noupdate)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/tools/convert.py", line 811, in convert_xml_import
    obj.parse(doc.getroot())
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/tools/convert.py", line 731, in parse
    self._tag_root(de)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/tools/convert.py", line 680, in _tag_root
    f(rec)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/tools/convert.py", line 693, in _tag_root
    raise ParseError('while parsing %s:%s, somewhere inside\n%s' % (
odoo.tools.convert.ParseError: while parsing /media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/data/crm_lead_demo.xml:339, somewhere inside
<function model="crm.lead" name="action_set_lost" eval="[[ref('crm_case_7'), ref('crm_case_9'), ref('crm_case_11'), ref('crm_case_12')]]" context="{'install_mode': True}"/>
2021-11-08 05:47:30,850 28289 WARNING odoo15-local odoo.modules.loading: Transient module states were reset 
2021-11-08 05:47:30,853 28289 ERROR odoo15-local odoo.modules.registry: Failed to load registry 
2021-11-08 05:47:30,854 28289 CRITICAL odoo15-local odoo.service.server: Failed to initialize database `odoo15-local`. 
Traceback (most recent call last):
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/service/server.py", line 1246, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/modules/registry.py", line 87, in new
    odoo.modules.load_modules(registry, force_demo, status, update_module)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/modules/loading.py", line 474, in load_modules
    processed_modules += load_marked_modules(cr, graph,
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/modules/loading.py", line 363, in load_marked_modules
    loaded, processed = load_module_graph(
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/modules/loading.py", line 249, in load_module_graph
    cr.commit()
  File "<decorator-gen-7>", line 2, in commit
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/sql_db.py", line 89, in check
    return f(self, *args, **kwargs)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/sql_db.py", line 466, in commit
    self.flush()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/sql_db.py", line 108, in flush
    self.transaction.flush()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/api.py", line 826, in flush
    env_to_flush['base'].flush()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 5622, in flush
    self.recompute()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 6095, in recompute
    process(field)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 6079, in process
    field.recompute(recs)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1243, in recompute
    self.compute_value(recs)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1265, in compute_value
    records._compute_field_value(self)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/mail/models/mail_thread.py", line 410, in _compute_field_value
    return super()._compute_field_value(field)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 4248, in _compute_field_value
    getattr(self, field.compute)()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 462, in _compute_prorated_revenue
    lead.prorated_revenue = round((lead.expected_revenue or 0.0) * (lead.probability or 0) / 100.0, 2)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1054, in __get__
    self.recompute(record)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1243, in recompute
    self.compute_value(recs)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1265, in compute_value
    records._compute_field_value(self)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/mail/models/mail_thread.py", line 410, in _compute_field_value
    return super()._compute_field_value(field)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 4248, in _compute_field_value
    getattr(self, field.compute)()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 451, in _compute_probabilities
    lead_probabilities = self._pls_get_naive_bayes_probabilities()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 1873, in _pls_get_naive_bayes_probabilities
    leads_values_dict = self._pls_get_lead_pls_values(domain=domain)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 2418, in _pls_get_lead_pls_values
    value = lead[field].id if isinstance(lead[field], models.BaseModel) else lead[field]
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 5866, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1054, in __get__
    self.recompute(record)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1243, in recompute
    self.compute_value(recs)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1265, in compute_value
    records._compute_field_value(self)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/mail/models/mail_thread.py", line 410, in _compute_field_value
    return super()._compute_field_value(field)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 4248, in _compute_field_value
    getattr(self, field.compute)()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 421, in _compute_phone_state
    if lead.phone:
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1054, in __get__
    self.recompute(record)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1243, in recompute
    self.compute_value(recs)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/fields.py", line 1265, in compute_value
    records._compute_field_value(self)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/mail/models/mail_thread.py", line 410, in _compute_field_value
    return super()._compute_field_value(field)
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/odoo/models.py", line 4248, in _compute_field_value
    getattr(self, field.compute)()
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 409, in _compute_phone
    if lead.partner_id.phone and lead._get_partner_phone_update():
  File "/media/hd2/prjs/odoo/odoo14.so/odoo/addons/crm/models/crm_lead.py", line 637, in _get_partner_phone_update
    partner_phone_formatted = self.partner_id.phone_get_sanitized_number(number_fname='phone') or self.partner_id.phone or False
AttributeError: 'res.partner' object has no attribute 'phone_get_sanitized_number'

```

### PR DESCRIPTION
**Pull request by amh-mw**
Fixes #79460.

### CODE REVIEW

**Comment by amh-mw** [comment]
The `auto_install` daisy chain causes the `sms` module to be installed. This gave me a minor heart attack a while back when I saw one of my Odoo 13 staging servers (with real customer phone numbers) attempting to send text messages. Given that we do not use `sms` or any of the other In-App Purchases modules, it is unhelpful for `iap` to be installed by default. Additionally, it provokes some privacy concerns about our customers phone numbers being sent to a third-party unexpectedly.

```
web_1      | 2020-10-20 18:09:13,864 21 INFO odoo odoo.addons.iap.models.iap: iap jsonrpc https://iap-sms.odoo.com/iap/sms/1/send 
web_1      | 2020-10-20 18:09:15,102 21 INFO odoo odoo.addons.sms.models.sms_sms: Send batch 1 SMS: [1]: gave [{'state': 'insufficient_credit', 'res_id': 1, 'credit': 0}] 
```

**Comment by MiquelRForgeFlow** [comment]
@tde-banana-odoo this PR is interesting

**Comment by tde-banana-odoo** [comment]
Hello,

Thanks for the ping ! Interesting feedback, especially I know phone / sms integration still has some glitches like this one. It seems we have different things here

  * indeed we should be able to use crm without IAP (and SMS);
  * we cannot add an inherit in stable that way in stable as it impacts the db (new fields notably, but anyway a functional change);
  * having IAP as auto_install is a choice done by Odoo to automatically propose IAP services by default;

I understand you don't want to deal with IAP or having to mock its calls by default. However this is not likely to change and auto install will probably stay. However you should be able to use everything after removing IAP. Seems there are some calls to mail.thread.phone methods that are not available; that seems to be the real fix to do (aka in 15: remove IAP and everything should work). It seems we would have to define ``phone_get_sanitized_numbers`` on partner directly in ``phone_validation``. Would that solve the issue ?

Cheers,

**Comment by amh-mw** [comment]
I need to run some more tests locally, but I think this is the minimum viable change for 15. I have split off a second pull request for 16 to solve this same problem with the inheritance change rather than copy pasta.

**Comment by MiquelRForgeFlow** [comment]
Second PR: https://github.com/odoo/odoo/pull/85313 (to keep tracking)

**Comment by amh-mw** [comment]
Copy pasta error fixed. All Odoo tests are passing for the modules we use, with `iam` never installed (via local patch setting auto_install to False).

File: addons/phone_validation/models/res_partner.py [hunk 1]
```diff
@@ -30,3 +36,15 @@ def _phone_format(self, number, country=None, company=None):
...
         )
+
+    # FIXME DRY
+    def _phone_get_country_field(self):
+        if 'country_id' in self:
+            return 'country_id'
+        return False
+
+    # OVERRIDE
+    def _phone_get_number_fields(self):
```
Comment (tde-banana-odoo) [inline addons/phone_validation/models/res_partner.py:47]: Is it used in any CRM flow without sms installed ?

File: addons/phone_validation/models/res_partner.py [hunk 2]
```diff
@@ -30,3 +36,15 @@ def _phone_format(self, number, country=None, company=None):
             force_format='INTERNATIONAL',
             raise_exception=False
         )
+
+    # FIXME DRY
+    def _phone_get_country_field(self):
+        if 'country_id' in self:
```
Comment (tde-banana-odoo) [inline addons/phone_validation/models/res_partner.py:42]: Actually as we know it is available in base, we could directly return 'country_id'

File: addons/phone_validation/models/res_partner.py [hunk 3]
```diff
@@ -19,6 +18,13 @@ def _onchange_mobile_validation(self):
         if self.mobile:
             self.mobile = self._phone_format(self.mobile)
 
+    # FIXME DRY
+    def phone_get_sanitized_number(self, number_fname='mobile', force_format='E164'):
```
Comment (tde-banana-odoo) [inline addons/phone_validation/models/res_partner.py:22]: Could be great to add a small docstring explaining why it is here

**Review by tde-banana-odoo — COMMENTED** [review]
Some review :)

File: addons/phone_validation/models/res_partner.py (hunk 1 above)
Comment (amh-mw) [reply addons/phone_validation/models/res_partner.py:47]: Our clients have call centers that would like to validate phone numbers, but do not text their customers.

**Comment by tde-banana-odoo** [comment]
Hello,

I updated the patch to be a bit more concise. It will be tested internally soon :) .

Cheers,

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-ff

**Pull request closed without merging**